File: herencia-dto/HerenciaDto-Java-IvanRC/src/main/java/storage/personas/PersonaStorageServiceJson.java
package storage.personas;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import config.ConfigApp;
import dto.ListaPersonasDto;
import mapper.PersonaMapper;
import models.Persona;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class PersonaStorageServiceJson implements PersonasStorageService{

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"personas.json");

    Moshi moshi = new Moshi.Builder().build();
    JsonAdapter<ListaPersonasDto> jsonAdapter = moshi.adapter(ListaPersonasDto.class);

    public PersonaStorageServiceJson() throws IOException {}

    @Override
    public void saveAll(List<Persona> entities) {
        try(BufferedWriter bw = new BufferedWriter(new FileWriter(file))){
            bw.write((jsonAdapter.indent("   ").toJson(PersonaMapper.toPersonasDto(entities))));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<Persona> loadAll() {
        List<Persona> personas = new ArrayList<>();
        if(!file.exists()) return personas;
        try{
            String json = Files.readString(Path.of(file.getPath()));
            personas = PersonaMapper.toPersonas(jsonAdapter.fromJson(json));
        }catch (IOException e){
            System.out.println("Error de lectura en ingrediente.json.");
        }
        return personas;
    }
}

File: herencia-dto/HerenciaDto-Kotlin-IvanRC/src/main/kotlin/storage/persona/PersonaStorageServiceJson.kt
package storage.persona

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.ConfigApp
import dto.ListaPersonasDto
import mapper.toPersonas
import mapper.toPersonasDto
import models.Persona
import mu.KotlinLogging
import java.io.File

@ExperimentalStdlibApi
class PersonaStorageServiceJson: PersonaStorageService {

    private val configApp = ConfigApp
    private val file = File(configApp.APP_DATA+File.separator+"personas.json")
    private val adapter = Moshi.Builder().addLast(KotlinJsonAdapterFactory()).build().adapter<ListaPersonasDto>()

    private val logger = KotlinLogging.logger {  }

    override fun saveAll(entities: List<Persona>) {
        logger.debug { "Se guardan todas las personas en el fichero JSON" }
        file.writeText(adapter.indent("   ").toJson(entities.toPersonasDto()))
    }

    override fun loadAll(): List<Persona> {
        logger.debug { "Se cargan todas las personas del fichero JSON" }
        if(!file.exists()) return emptyList()
        return adapter.fromJson(file.readText())!!.toPersonas()
    }
}